src/main/credentialServer/httpCredentialServer.ts

```diff
@@ -0,0 +1,210 @@
+import fs from 'fs';
+import http from 'http';
+import path from 'path';
+import { URL } from 'url';
+import childProcess from 'child_process';
+
+import Logging from '@/utils/logging';
+import paths from '@/utils/paths';
+import * as serverHelper from '@/main/serverHelper';
+import { findHomeDir } from '@/config/findHomeDir';
+
+export type ServerState = {
+  user: string;
+  password: string;
+  port: number;
+  pid: number;
+}
+
+const console = Logging.server;
+const SERVER_PORT = 6109;
+const SERVER_USERNAME = 'user';
+const SERVER_FILE_BASENAME = 'credential-server.json';
+const MAX_REQUEST_BODY_LENGTH = 2048;
+
+type dispatchFunctionType = (helperName: string, data: string, request: http.IncomingMessage, response: http.ServerResponse) => Promise<void>;
+
+type checkFunctionOutputType = (stdout: string, stderr: string) => boolean;
+
+function requireNoOutput(stdout: string, stderr: string): boolean {
+  return !stdout && !stderr;
+}
+
+function requireJSONOutput(stdout: string, stderr: string): boolean {
+  if (stderr) {
+    return false;
+  }
+  try {
+    JSON.parse(stdout);
+
+    return true;
+  } catch {}
+
+  return false;
+}
+
+export class HttpCredentialServer {
```

Comment:
> (If we plan to use vsock later, shouldn't this be written in golang instead of TypeScript? This can't be over TCP permanently due to security considerations and port collisions, right?)

I'm not sure if this has to go over vsock; we should always be able to connect to the host over TCP from inside the VM. And the server should pick a random free port (maybe not yet implemented?), so that is not an issue either.

The server should only accept connections from localhost and the VM, but also requires a token/secret, so this should be reasonably secure: traffic never leaves the host, so should be un-snoopable. Anyone who can impersonate the current user on the local system has access to the credentials anyways, so that is not an attack vector.

Maybe dropping of outside connections isn't implemented either, but it also doesn't matter, as the auth secret should still protect the endpoint.

What advantage would we get from tunnelling over vsock? I know I put this into the epic, but I wonder if there is a reason to do this.

Reply:
I've made the name change. Looks like that's all this comment-sub-thread requires.